File: data/history_data/ParaEs_Synchronous/pv_plot/pv_plot.py
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt

input_data = pd.read_csv('../../data/1011PVtest.csv')
fnamevar = "../../variables_test2pv1014.dat"

input_data.columns = ['Time', 'V', 'I', 'Is']
input_data.drop(columns='Is', inplace=True)
input_data.set_index('Time', inplace=True)

step_time = input_data.index[1] - input_data.index[0]
dI_dt = (input_data.I.to_numpy()[2:] - input_data.I.to_numpy()[:-2]) / step_time / 2
dV_dt = (input_data.V.to_numpy()[2:] - input_data.V.to_numpy()[:-2]) / step_time / 2

lines = open(fnamevar, "r").readlines()
# read output data in fnamevar (this line is a long story...)
Chat = np.array(
    [
        np.fromstring(
            min(re.findall(re.escape("[") + "(.*?)" + re.escape("]"), line), key=len),
            sep=",",
        )
        for line in lines
    ]
)

# Kt, Rs, Rsh, Iso
Kt = Chat[:, 0]
Rs = Chat[:, 1]
Rsh = Chat[:, 2] * 1e4
Iso = Chat[:, 3]

left = 1 + Iso*Rs/Kt*np.exp((np.repeat(np.reshape(input_data.V.to_numpy()[1:-1],(-1,1)), Chat.shape[0], axis =1)+np.outer(input_data.I.to_numpy()[1:-1], Rs))/Kt)  + Rs/Rsh

right = - np.repeat(np.reshape(dV_dt,(-1,1)), Chat.shape[0], axis =1) * (Iso/Kt*np.exp((np.repeat(np.reshape(input_data.V.to_numpy()[1:-1],(-1,1)), Chat.shape[0], axis =1)+np.outer(input_data.I.to_numpy()[1:-1], Rs))/Kt) + 1/Rsh)

result = left*np.repeat(np.reshape(dI_dt,(-1,1)), Chat.shape[0], axis =1) - right

plt.plot(np.mean(np.abs(result), axis=0))

plt.figure()
plt.plot(np.mean(np.abs(result), axis=0))
plt.plot(range(Chat.shape[0]), np.ones(Chat.shape[0]) * 0, "k--")
plt.ylim((-0.1, 1))
plt.show()

plt.figure()
plt.plot(np.mean(np.abs(result), axis=0))
plt.plot(range(Chat.shape[0]), np.ones(Chat.shape[0]) * 0, "k--")
plt.xlim((0,200))
plt.show()

l, c = Chat.shape
plt.plot(range(l), Chat[:, 0], color="#10739e", linestyle='--')
plt.plot(range(l), Chat[:, 1], color="#f2931e", linestyle='--')
plt.plot(range(l), Chat[:, 2], color="#ae4132", linestyle='--')
plt.plot(range(l), Chat[:, 3], color="#23445d", linestyle='--')
# plt.plot(range(l), np.ones(Chat[:, 0].shape) * 1.5, "r--")
# plt.plot(range(l), np.ones(Chat[:, 1].shape) * 0.15, "b--")
plt.legend(["$K_t$","$R_{s}$","$R_{sh}$","$I_so$"], loc="right")
plt.xlabel("Epoch")
plt.show()

File: data/history_data/ParaEs_Synchronous/pv_plot/pvplot.m
% Load data
input_data = readtable('data/1011PVtest.csv');
fnamevar = "variables_testpv1014.dat";

% Rename and drop columns
input_data.Properties.VariableNames = {'Time', 'V', 'I', 'Is'};
input_data.Is = [];

% Calculate derivatives
step_time = input_data.Time(2) - input_data.Time(1);
dI_dt = (input_data.I(3:end) - input_data.I(1:end-2)) / step_time / 2;
dV_dt = (input_data.V(3:end) - input_data.V(1:end-2)) / step_time / 2;

% Read the output data in fnamevar
fid = fopen(fnamevar, 'r');
lines = textscan(fid, '%s', 'Delimiter', '\n');
lines = lines{1};
fclose(fid);

Chat = [];
for i = 1:length(lines)
    match = regexp(lines{i}, '\[(.*?)\]', 'tokens');
    if ~isempty(match)
        line_data = str2double(strsplit(match{1}{1}, ','));
        Chat = [Chat; line_data];
    end
end

% Extract variables
Kt = Chat(:, 1);
Rs = Chat(:, 2);
Rsh = Chat(:, 3) * 1e4;
Iso = Chat(:, 4);

V_mat = repmat(input_data.V(2:end-1), 1, length(Kt));
I_mat = repmat(input_data.I(2:end-1), 1, length(Kt));
dI_dt_mat = repmat(dI_dt, 1, length(Kt));
dV_dt_mat = repmat(dV_dt, 1, length(Kt));

Rs_mat = repmat(Rs', size(I_mat, 1), 1);
Kt_mat = repmat(Kt', size(I_mat, 1), 1);
Rsh_mat = repmat(Rsh', size(I_mat, 1), 1);

left = 1 + (Iso' .* Rs' ./ Kt') .* exp((V_mat + I_mat .* Rs_mat) ./ Kt_mat) + Rs_mat ./ Rsh_mat;
right = -dV_dt_mat .* ((Iso' ./ Kt_mat) .* exp((V_mat + I_mat .* Rs_mat) ./ Kt_mat) + 1 ./ Rsh_mat);
result = left .* dI_dt

% Plotting results
figure;
plot(mean(abs(result), 1));
title('Mean of Absolute Value of Result');

figure;
plot(mean(abs(result), 1));
hold on;
plot(1:size(Chat, 1), ones(1, size(Chat, 1)) * 0, "k--");
ylim([-0.1, 1]);
title('Mean of Absolute Value of Result with Horizontal Line');

figure;
plot(mean(abs(result), 1));
hold on;
plot(1:size(Chat, 1), ones(1, size(Chat, 1)) * 0, "k--");
xlim([0, 1000]);
title('Zoomed-in View of Previous Plot');
